src/core/filter-from-graph.ts

```diff
@@ -0,0 +1,98 @@
+import { DepGraph, DepGraphInternal, NodeInfo, PkgInfo } from './types';
+import { DepGraphBuilder } from './builder';
+import * as isEmpty from 'lodash.isempty';
+
+type NodeId = string;
+type PkgName = string;
+
+export async function filterPackagesFromGraph(
+  baseDepGraph: DepGraph,
+  packagesToFilterOut: (PkgName | PkgInfo)[],
+): Promise<DepGraph> {
+  if (!packagesToFilterOut?.length) return baseDepGraph;
+
+  const depGraph = baseDepGraph as DepGraphInternal;
+
+  const packages = depGraph
+    .getDepPkgs()
+    .filter((existingPkg) =>
+      packagesToFilterOut.some((pkgToFilter) =>
+        isString(pkgToFilter)
+          ? existingPkg.name === pkgToFilter
+          : existingPkg.name === pkgToFilter.name &&
+            existingPkg.version === pkgToFilter.version,
+      ),
+    );
+
+  const nodeIdsToFilterOut: NodeId[] = [];
+  for (const pkg of packages) {
+    const nodeIds = depGraph.getPkgNodeIds(pkg);
+    nodeIdsToFilterOut.push(...nodeIds);
+  }
+
+  return filterNodesFromGraph(baseDepGraph, nodeIdsToFilterOut);
+}
+
+export async function filterNodesFromGraph(
+  baseDepGraph: DepGraph,
+  nodeIdsToFilterOut: NodeId[],
+): Promise<DepGraph> {
+  if (!nodeIdsToFilterOut?.length) return baseDepGraph;
+
+  const depGraph = baseDepGraph as DepGraphInternal;
+  const existingNodeIds = new Set(depGraph['_graph'].nodes());
+  nodeIdsToFilterOut = nodeIdsToFilterOut.filter((nodeId) =>
+    existingNodeIds.has(nodeId),
+  );
+  if (nodeIdsToFilterOut.length === 0) return baseDepGraph;
+
+  const depGraphBuilder = new DepGraphBuilder(
+    depGraph.pkgManager,
+    depGraph.rootPkg,
+  );
+
+  const nodeIdsToFilterOutSet = new Set(nodeIdsToFilterOut);
+
+  dfsBuildGraph(
+    depGraph.rootNodeId,
+    depGraph,
+    depGraphBuilder,
+    nodeIdsToFilterOutSet,
+  );
```

Comment:
Oof... not a fan of this side effect style builder. Hard to keep track of the flow.